prov/efa/src/efa_hmem.c

```diff
@@ -79,266 +107,162 @@ static int efa_domain_hmem_info_init_protocol_thresholds(struct efa_domain *efa_
  *
  * @param[in,out]  efa_domain  Pointer to struct efa_domain
+ * @param[in]      iface       HMEM interface
  *
  * @return         0
  */
-static int efa_domain_hmem_info_init_synapseai(struct efa_domain *efa_domain)
+static int
+efa_domain_hmem_info_init_iface(struct efa_domain *efa_domain, enum fi_hmem_iface iface)
 {
-#if HAVE_SYNAPSEAI
-	size_t tmp_value;
-	struct efa_hmem_info *info = &efa_domain->hmem_info[FI_HMEM_SYNAPSEAI];
+	struct efa_hmem_info *info = &efa_domain->hmem_info[iface];
 
-	if (!ofi_hmem_is_initialized(FI_HMEM_SYNAPSEAI)) {
-		EFA_INFO(FI_LOG_DOMAIN, "FI_HMEM_SYNAPSEAI is not initialized\n");
+	if (!ofi_hmem_is_initialized(iface)) {
+		EFA_INFO(FI_LOG_DOMAIN, "%s is not initialized\n",
+		         fi_tostr(&iface, FI_TYPE_HMEM_IFACE));
 		return 0;
 	}
 
-	if (!(g_device_list[0].device_caps & EFADV_DEVICE_ATTR_CAPS_RDMA_READ)) {
-		EFA_WARN(FI_LOG_DOMAIN,
-			 "No EFA RDMA read support, transfers using Habana Gaudi will fail.\n");
-		return 0;
+	info->p2p_disabled_by_user = false;
+	info->p2p_supported_by_device = efa_domain_hmem_support_p2p(efa_domain, iface);
+	if (!info->p2p_supported_by_device) {
+		EFA_INFO(FI_LOG_DOMAIN, "%s P2P support is not available.\n",
+		         fi_tostr(&iface, FI_TYPE_HMEM_IFACE));
 	}
 
-	info->initialized = true;
-	info->p2p_disabled_by_user = false;
-	/* SynapseAI currently requires P2P */
-	info->p2p_required_by_impl = true;
-	info->p2p_supported_by_device = true;
-	efa_domain_hmem_info_init_protocol_thresholds(efa_domain, FI_HMEM_SYNAPSEAI);
-
-	/*  Only the long read protocol is supported */
-	if (-FI_ENODATA != fi_param_get_size_t(&efa_prov, "inter_max_medium_message_size", &tmp_value) ||
-		-FI_ENODATA != fi_param_get_size_t(&efa_prov, "inter_min_read_message_size", &tmp_value) ||
-		-FI_ENODATA != fi_param_get_size_t(&efa_prov, "inter_min_read_write_size", &tmp_value) ||
-		-FI_ENODATA != fi_param_get_size_t(&efa_prov, "runt_size", &tmp_value)) {
-		EFA_WARN(FI_LOG_DOMAIN,
-				"One or more of the following environment variable(s) were set: ["
-				"FI_EFA_INTER_MAX_MEDIUM_MESSAGE_SIZE, "
-				"FI_EFA_INTER_MIN_READ_MESSAGE_SIZE, "
-				"FI_EFA_INTER_MIN_READ_WRITE_SIZE, "
-				"FI_EFA_RUNT_SIZE"
-				"], but EFA HMEM via Synapse only supports long read protocol. "
-				"The variable(s) will not modify Synapse memory run config.\n");
+	switch (iface) {
+	case FI_HMEM_CUDA:
+		/* If user is using libfabric API 1.18 or later, by default EFA
+		 * provider is permitted to use CUDA library to support CUDA
+		 * memory, therefore p2p is not required.
+		 */
+		if (FI_VERSION_GE(efa_domain->util_domain.fabric->fabric_fid.api_version,
+				  FI_VERSION(1, 18)))
+			info->p2p_required_by_impl = !hmem_ops[iface].initialized;
+		else
+			info->p2p_required_by_impl = true;
+		break;
+	case FI_HMEM_NEURON:
+		info->p2p_required_by_impl = true;
+		break;
+	case FI_HMEM_SYNAPSEAI:
+		if (!(g_device_list[0].device_caps & EFADV_DEVICE_ATTR_CAPS_RDMA_READ)) {
+			EFA_WARN(FI_LOG_DOMAIN,
```

Comment:
I think we can remove this check now, because we already check this during FI_HMEM check in fi_getinfo, see https://github.com/ofiwg/libfabric/blob/main/prov/efa/src/efa_user_info.c#L312-L319 . If rdma-read is not enabled, we won't support FI_HMEM unless user request libfabric 1.18 API and it has cuda interface initialized

Reply:
Makes sense. Will update.